crates/ra_hir/src/ty/infer.rs

```diff
@@ -967,8 +988,51 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
                 Ty::unit()
             }
             Expr::For { iterable, body, pat } => {
-                let _iterable_ty = self.infer_expr(*iterable, &Expectation::none());
-                self.infer_pat(*pat, &Ty::Unknown, BindingMode::default());
+                use crate::nameres::PerNs;
+                use crate::resolve::Resolution::Def;
+                use crate::code_model::ModuleDef::Trait;
+                use super::traits::ProjectionPredicate;
+                use super::ProjectionTy;
+
+                let iterable_ty = self.infer_expr(*iterable, &Expectation::none());
+
+                let into_iter_path = Path {
+                    kind: PathKind::Plain,
+                    segments: vec![
+                        PathSegment { name: Name::from_str("std"), args_and_bindings: None },
+                        PathSegment { name: Name::from_str("iter"), args_and_bindings: None },
+                        PathSegment { name: Name::from_str("IntoIterator"), args_and_bindings: None },
```

Comment:
@matklad do you have an idea how to handle these 'well-known' symbols? there's going to be a lot more of these in the future...

Reply:
Yeah, we definitely do not want to expose `Name`'s constructor due to hygiene which we'll implement some day. 

We already have `KnownName` `enum`, so the simplest thing to do would be to add more variants to it. 

I think I figured a way to make `SmolStr::new` a const-fn, so we will be able to refactor this to a better solution soon:

```
mod known_paths {
    const STD_ITERATOR: Path = ...
}
```

Reply:
https://github.com/rust-analyzer/smol_str/commit/c7ce079a24cdd67eaf972f070874c6be7c166082#diff-b4aea3e418ccdb71239b96952d9cddb6R26

![image](https://user-images.githubusercontent.com/1711539/60767378-d4bf1100-a0bf-11e9-9767-55f9df218c40.png)

Reply:
So for this change, should I introduce the path as KnownName?

Reply:
@unrealhoang yep, specifically, you should ad std, iter and IntoIterator as known names, remove, `SmolStr::from_str` constructor and construct paths as

```
PathSegment { name: KnownName::Std.as_name(), args_and_bindings: None },
```

Reply:
Filed https://github.com/rust-analyzer/rust-analyzer/issues/1503 for that KnownName refactoring